src/common/src/metrics/guarded_metrics.rs

```diff
@@ -94,40 +102,60 @@ pub type LabelGuardedIntCounterVec<const N: usize> =
 pub type LabelGuardedIntGaugeVec<const N: usize> =
     LabelGuardedMetricVec<VecBuilderOfGauge<AtomicI64>, N>;
 
-pub type LabelGuardedHistogram = LabelGuardedMetric<VecBuilderOfHistogram>;
-pub type LabelGuardedIntCounter = LabelGuardedMetric<VecBuilderOfCounter<AtomicU64>>;
-pub type LabelGuardedIntGauge = LabelGuardedMetric<VecBuilderOfGauge<AtomicI64>>;
+pub type LabelGuardedHistogram<const N: usize> = LabelGuardedMetric<VecBuilderOfHistogram, N>;
+pub type LabelGuardedIntCounter<const N: usize> =
+    LabelGuardedMetric<VecBuilderOfCounter<AtomicU64>, N>;
+pub type LabelGuardedIntGauge<const N: usize> = LabelGuardedMetric<VecBuilderOfGauge<AtomicI64>, N>;
 
 fn gen_test_label<const N: usize>() -> [&'static str; N] {
-    vec!["test"; N].try_into().unwrap()
+    const TEST_LABELS: [&str; 5] = ["test1", "test2", "test3", "test4", "test5"];
+    (0..N)
+        .map(|i| TEST_LABELS[i])
+        .collect_vec()
+        .try_into()
+        .unwrap()
 }
 
 #[derive(Clone)]
 pub struct LabelGuardedMetricVec<T: MetricVecBuilder, const N: usize> {
     inner: MetricVec<T>,
-    _labels: [&'static str; N],
+    labeled_metrics_count: Arc<Mutex<HashMap<[String; N], usize>>>,
+    labels: [&'static str; N],
+}
+
+impl<T: MetricVecBuilder, const N: usize> Debug for LabelGuardedMetricVec<T, N> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct(format!("LabelGuardedMetricVec<{}, {}>", type_name::<T>(), N).as_str())
+            .field("label", &self.labels)
+            .finish()
+    }
 }
 
 impl<T: MetricVecBuilder, const N: usize> LabelGuardedMetricVec<T, N> {
     pub fn new(inner: MetricVec<T>, labels: &[&'static str; N]) -> Self {
         Self {
             inner,
-            _labels: *labels,
+            labeled_metrics_count: Default::default(),
+            labels: *labels,
         }
     }
 
-    pub fn with_label_values(&self, labels: &[&str; N]) -> LabelGuardedMetric<T> {
+    pub fn with_label_values(&self, labels: &[&str; N]) -> LabelGuardedMetric<T, N> {
+        let mut count_guard = self.labeled_metrics_count.lock();
+        let label_string = labels.map(|str| str.to_string());
+        *count_guard.entry(label_string).or_insert(0) += 1;
```

Comment:
Well, it's sad to find that there's very similar logic in the `prometheus` crate: there's also `HashMap` and `Arc` to deduplicate the metric struct for the same labels. We're actually paying an unnecessary cost for its interface design.

Reply:
Indeed. In the best case, the `prometheus` should do the reference counting by itself, and we shouldn't care about removing the labels🥹